File: web-ui/backend/app/services/jwt_service.rb
class JwtService
  SECRET = ENV.fetch("WEB_SECRET_KEY", "change-me-in-production")
  EXPIRE_MINUTES = ENV.fetch("WEB_JWT_EXPIRE_MINUTES", "1440").to_i

  def self.encode(canonical_user_id, extra_claims = {})
    payload = {
      sub: canonical_user_id,
      exp: EXPIRE_MINUTES.minutes.from_now.to_i,
      iat: Time.now.to_i
    }.merge(extra_claims)
    JWT.encode(payload, SECRET, "HS256")
  end

  def self.decode(token)
    JWT.decode(token, SECRET, true, algorithm: "HS256").first
  rescue JWT::DecodeError, JWT::ExpiredSignature, JWT::VerificationError
    nil
  end
end

File: services/identity/jwt_service.py
"""JWT signing and verification."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone

import jwt

from identity.config import JWT_ALGORITHM, JWT_EXPIRE_MINUTES, JWT_SECRET


def encode(canonical_user_id: str, name: str = "") -> str:
    payload = {
        "sub": canonical_user_id,
        "name": name,
        "iat": datetime.now(timezone.utc),
        "exp": datetime.now(timezone.utc) + timedelta(minutes=JWT_EXPIRE_MINUTES),
    }
    return jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)


def decode(token: str) -> dict | None:
    try:
        return jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
    except (jwt.ExpiredSignatureError, jwt.InvalidTokenError):
        return None
